persistence/scaladsl/src/main/scala/com/lightbend/lagom/scaladsl/persistence/PersistentEntity.scala

```diff
@@ -142,114 +133,78 @@ abstract class PersistentEntity[Command, Event, State] {
+        case params @ (_, ctx, _) if handler.isDefinedAt(params) =>
+          handler(params)
+          ctx.done
+      }
 
-  }
+      new Behavior(state, eventHandler, commandHandler.orElse(delegate))
+    }
 
-  /**
-   * Behavior consists of current state and functions to process incoming commands
-   * and persisted events. `Behavior` is an immutable class. Use the [[BehaviorBuilder]]
-   * for defining command and event handlers.
-   */
-  final case class Behavior(
-    state:          State,
-    eventHandler:   Function[Event, Option[Behavior]],
-    commandHandler: (Command, CommandContext[Any]) => Option[Persist[_ <: Event]]
-  ) {
+    def addEventHandler(handler: EventHandler) = {
+      new Behavior(state, eventHandler.orElse(handler), commandHandler)
+    }
 
     /**
      * @return new instance with the given state
      */
     def withState(newState: State): Behavior =
-      copy(state = newState)
-
-    def transformState(f: State => State): Behavior =
-      copy(state = f(state))
-  }
-
-  /**
-   * Builder that is used for defining the event and command handlers.
-   * Use [#build] to create the immutable [[Behavior]].
-   */
-  protected final class BehaviorBuilder(
-    val state:       State,
-    val cmdHandlers: PartialFunction[Command, Function[CommandContext[Any], Option[Persist[_ <: Event]]]] = Map.empty,
-    val evtHandlers: PartialFunction[Event, Behavior]                                                     = PartialFunction.empty
-  ) {
-
-    // TODO apidocs
-    def withState(state: State): BehaviorBuilder =
-      copy(state = state)
-
-    // TODO apidocs
-    def withEventHandlers(evtHandlers: PartialFunction[Any, Behavior]): BehaviorBuilder =
-      copy(evtHandlers = evtHandlers)
-
-    // TODO apidocs
-    def withCommandHandlers(cmdHandlers: PartialFunction[Any, Function[CommandContext[Any], Option[Persist[_ <: Event]]]]): BehaviorBuilder =
-      copy(cmdHandlers = cmdHandlers)
-
-    private def copy(
-      state:       State                                                                                = state,
-      evtHandlers: PartialFunction[Event, Behavior]                                                     = evtHandlers,
-      cmdHandlers: PartialFunction[Command, Function[CommandContext[Any], Option[Persist[_ <: Event]]]] = cmdHandlers
-    ) =
-      new BehaviorBuilder(state, cmdHandlers, evtHandlers)
+      new Behavior(state = newState, eventHandler, commandHandler)
 
     /**
-     * Construct the corresponding immutable `Behavior`.
+     * Transform the state
      */
-    def build(): Behavior = {
-
-      def handler(cmd: Command, ctx: CommandContext[Any]): Option[Persist[_ <: Event]] = {
-        (cmdHandlers.orElse[Command, Function[CommandContext[Any], Option[Persist[_ <: Event]]]]({ case _: Command => (ctx: CommandContext[Any]) => None }))(cmd)(ctx)
-      }
+    def map(f: State => State): Behavior =
+      new Behavior(state = f(state), eventHandler, commandHandler)
```

Comment:
I wonder if it would be clearer if this was called `mapState`. Just thinking about the readability of this:

```scala
case (Appended(elem), b)  => b.map(_.add(elem))
```
vs

```scala
case (Appended(elem), b)  => b.mapState(_.add(elem))
```

It makes it clear what is actually being mapped, whereas the first it's not obvious - are we mapping an option?  What is b? We also need to take into account that changing state in this way makes behaviour a fundamental concept - with the current Java API, you can use persistent entities without ever changing behaviour, and changing behaviour is almost considered an advanced feature.  This makes for a more gentle introduction, you can first learn how to use persistent entities with just one behaviour, and then learn how to work with multiple behaviours.  But this API does require teaching behaviours straight up, since that's the primary thing that's interacted with.  Not saying that's necessarily a bad thing, but we do need to consider it.